In [1]:
import pandas as pd
import numpy as np

In [3]:
food_info = pd.read_csv("food_info.csv")
food_info.head()

NDB_No                 Shrt_Desc  Water_(g)  Energ_Kcal  Protein_(g)  \
0    1001          BUTTER WITH SALT      15.87         717         0.85   
1    1002  BUTTER WHIPPED WITH SALT      15.87         717         0.85   
2    1003      BUTTER OIL ANHYDROUS       0.24         876         0.28   
3    1004               CHEESE BLUE      42.41         353        21.40   
4    1005              CHEESE BRICK      41.11         371        23.24   

   Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  Sugar_Tot_(g)  \
0          81.11     2.11            0.06           0.0           0.06   
1          81.11     2.11            0.06           0.0           0.06   
2          99.48     0.00            0.00           0.0           0.00   
3          28.74     5.11            2.34           0.0           0.50   
4          29.68     3.18            2.79           0.0           0.51   

        ...        Vit_A_IU  Vit_A_RAE  Vit_E_(mg)  Vit_D_mcg  Vit_D_IU  \
0       ...          2499.0      684.0        2.32        1.5      60.0   
1       ...          2499.0      684.0        2.32        1.5      60.0   
2       ...          3069.0      840.0        2.80        1.8      73.0   
3       ...           721.0      198.0        0.25        0.5      21.0   
4       ...          1080.0      292.0        0.26        0.5      22.0   

   Vit_K_(mcg)  FA_Sat_(g)  FA_Mono_(g)  FA_Poly_(g)  Cholestrl_(mg)  
0          7.0      51.368       21.021        3.043           215.0  
1          7.0      50.489       23.426        3.012           219.0  
2          8.6      61.924       28.732        3.694           256.0  
3          2.4      18.669        7.778        0.800            75.0  
4          2.5      18.764        8.598        0.784            94.0  

[5 rows x 36 columns]

In [6]:
column_names = food_info.columns
column_names

Index(['NDB_No', 'Shrt_Desc', 'Water_(g)', 'Energ_Kcal', 'Protein_(g)',
       'Lipid_Tot_(g)', 'Ash_(g)', 'Carbohydrt_(g)', 'Fiber_TD_(g)',
       'Sugar_Tot_(g)', 'Calcium_(mg)', 'Iron_(mg)', 'Magnesium_(mg)',
       'Phosphorus_(mg)', 'Potassium_(mg)', 'Sodium_(mg)', 'Zinc_(mg)',
       'Copper_(mg)', 'Manganese_(mg)', 'Selenium_(mcg)', 'Vit_C_(mg)',
       'Thiamin_(mg)', 'Riboflavin_(mg)', 'Niacin_(mg)', 'Vit_B6_(mg)',
       'Vit_B12_(mcg)', 'Vit_A_IU', 'Vit_A_RAE', 'Vit_E_(mg)', 'Vit_D_mcg',
       'Vit_D_IU', 'Vit_K_(mcg)', 'FA_Sat_(g)', 'FA_Mono_(g)', 'FA_Poly_(g)',
       'Cholestrl_(mg)'],
      dtype='object')

In [7]:
food_info.shape

(8618, 36)

In [8]:
# finding out how many missing values I have
missing_values_count = food_info.isnull().sum()
missing_values_count

NDB_No                0
Shrt_Desc             0
Water_(g)             6
Energ_Kcal            0
Protein_(g)           0
Lipid_Tot_(g)         0
Ash_(g)             332
Carbohydrt_(g)        0
Fiber_TD_(g)        656
Sugar_Tot_(g)      1939
Calcium_(mg)        354
Iron_(mg)           147
Magnesium_(mg)      682
Phosphorus_(mg)     572
Potassium_(mg)      410
Sodium_(mg)          83
Zinc_(mg)           701
Copper_(mg)        1255
Manganese_(mg)     2140
Selenium_(mcg)     1750
Vit_C_(mg)          792
Thiamin_(mg)        679
Riboflavin_(mg)     657
Niacin_(mg)         681
Vit_B6_(mg)         941
Vit_B12_(mcg)      1191
Vit_A_IU            686
Vit_A_RAE          1529
Vit_E_(mg)         3005
Vit_D_mcg          3299
Vit_D_IU           3298
Vit_K_(mcg)        3649
FA_Sat_(g)          344
FA_Mono_(g)         671
FA_Poly_(g)         664
Cholestrl_(mg)      368
dtype: int64

In [9]:
total_cells = np.product(food_info.shape)
missing_cells = missing_values_count.sum()
# percentage of missing values
(missing_cells/total_cells)*100

10.79168922926175

Im missing almost 11% of my data, which isnt terrible.

In [10]:
food_info.tail()

NDB_No                   Shrt_Desc  Water_(g)  Energ_Kcal  Protein_(g)  \
8613   83110             MACKEREL SALTED      43.00         305        18.50   
8614   90240  SCALLOP (BAY&SEA) CKD STMD      70.25         111        20.54   
8615   90480                  SYRUP CANE      26.00         269         0.00   
8616   90560                   SNAIL RAW      79.20          90        16.10   
8617   93600            TURTLE GREEN RAW      78.50          89        19.80   

      Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  Sugar_Tot_(g)  \
8613          25.10    13.40            0.00           0.0            0.0   
8614           0.84     2.97            5.41           0.0            0.0   
8615           0.00     0.86           73.14           0.0           73.2   
8616           1.40     1.30            2.00           0.0            0.0   
8617           0.50     1.20            0.00           0.0            0.0   

           ...        Vit_A_IU  Vit_A_RAE  Vit_E_(mg)  Vit_D_mcg  Vit_D_IU  \
8613       ...           157.0       47.0        2.38       25.2    1006.0   
8614       ...             5.0        2.0        0.00        0.0       2.0   
8615       ...             0.0        0.0        0.00        0.0       0.0   
8616       ...           100.0       30.0        5.00        0.0       0.0   
8617       ...           100.0       30.0        0.50        0.0       0.0   

      Vit_K_(mcg)  FA_Sat_(g)  FA_Mono_(g)  FA_Poly_(g)  Cholestrl_(mg)  
8613          7.8       7.148        8.320        6.210            95.0  
8614          0.0       0.218        0.082        0.222            41.0  
8615          0.0       0.000        0.000        0.000             0.0  
8616          0.1       0.361        0.259        0.252            50.0  
8617          0.1       0.127        0.088        0.170            50.0  

[5 rows x 36 columns]

In [23]:
food_info.describe()

NDB_No    Water_(g)   Energ_Kcal  Protein_(g)  Lipid_Tot_(g)  \
count   8618.000000  8612.000000  8618.000000   8618.00000    8618.000000   
mean   15467.757368    54.163710   226.438617     11.52391      10.647024   
std     9221.428766    30.648283   169.388910     10.55059      15.866353   
min     1001.000000     0.000000     0.000000      0.00000       0.000000   
25%     8695.250000    30.087500    93.000000      2.47000       0.992500   
50%    14273.000000    63.265000   191.000000      8.28500       5.235000   
75%    20050.750000    77.610000   336.000000     19.97750      13.900000   
max    93600.000000   100.000000   902.000000     88.32000     100.000000   

           Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  Sugar_Tot_(g)  Calcium_(mg)  \
count  8286.000000     8618.000000   7962.000000    6679.000000   8264.000000   
mean      1.816169       21.819062      2.189940       8.464779     76.556026   
std       3.035264       27.239000      4.447017      14.920260    205.045513   
min       0.000000        0.000000      0.000000       0.000000      0.000000   
25%       0.860000        0.040000      0.000000       0.000000     10.000000   
50%       1.200000        8.945000      0.700000       1.690000     21.000000   
75%       2.140000       32.900000      2.500000       9.000000     67.000000   
max      99.800000      100.000000     79.000000      99.800000   7364.000000   

            ...             Vit_A_IU     Vit_A_RAE   Vit_E_(mg)    Vit_D_mcg  \
count       ...          7932.000000   7089.000000  5613.000000  5319.000000   
mean        ...           744.119516    114.236564     1.339343     0.578060   
std         ...          3871.474249    857.972785     4.703431     3.887389   
min         ...             0.000000      0.000000     0.000000     0.000000   
25%         ...             0.000000      0.000000     0.120000     0.000000   
50%         ...            33.000000      4.000000     0.300000     0.000000   
75%         ...           259.000000     34.000000     0.790000     0.200000   
max         ...        100000.000000  30000.000000   149.400000   250.000000   

           Vit_D_IU  Vit_K_(mcg)   FA_Sat_(g)  FA_Mono_(g)  FA_Poly_(g)  \
count   5320.000000  4969.000000  8274.000000   7947.00000  7954.000000   
mean      23.152820    15.671342     3.621685      4.14469     2.296772   
std      155.482812    81.221277     6.558377      7.05124     5.206410   
min        0.000000     0.000000     0.000000      0.00000     0.000000   
25%        0.000000     0.200000     0.240000      0.25650     0.230000   
50%        0.000000     1.600000     1.613500      1.98100     0.711000   
75%        7.000000     5.200000     4.410000      5.22000     2.090000   
max    10000.000000  1714.500000    95.600000     83.68900    74.623000   

       Cholestrl_(mg)  
count     8250.000000  
mean        41.457212  
std        120.700142  
min          0.000000  
25%          0.000000  
50%          5.000000  
75%         67.000000  
max       3100.000000  

[8 rows x 35 columns]

In [57]:
# indexing a column as a DataFrame
water = food_info[['Water_(g)']]

In [25]:
protein_lipid_carb = food_info[['Protein_(g)', 'Lipid_Tot_(g)', 'Carbohydrt_(g)']]

Lets say I want to select only the columns that use grams as their unit of measurement - "(g)".

In [26]:
columns = food_info.columns
columns_lst = columns.tolist()
gram_columns = []
for a in columns_lst:
    if a.endswith("(g)"):
        gram_columns.append(a)
gram_df = food_info[gram_columns]
print(gram_df.head(3))

   Water_(g)  Protein_(g)  Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  \
0      15.87         0.85          81.11     2.11            0.06   
1      15.87         0.85          81.11     2.11            0.06   
2       0.24         0.28          99.48     0.00            0.00   

   Fiber_TD_(g)  Sugar_Tot_(g)  FA_Sat_(g)  FA_Mono_(g)  FA_Poly_(g)  
0           0.0           0.06      51.368       21.021        3.043  
1           0.0           0.06      50.489       23.426        3.012  
2           0.0           0.00      61.924       28.732        3.694  


Or select all of the columns that use milligrams as their unit of measurement - "(mg)".

In [29]:
mg_columns = []
for a in columns_lst:
    if a.endswith("(mg)"):
        mg_columns.append(a)
mg_df = food_info[mg_columns]
print(mg_df.head(3))

   Calcium_(mg)  Iron_(mg)  Magnesium_(mg)  Phosphorus_(mg)  Potassium_(mg)  \
0          24.0       0.02             2.0             24.0            24.0   
1          24.0       0.16             2.0             23.0            26.0   
2           4.0       0.00             0.0              3.0             5.0   

   Sodium_(mg)  Zinc_(mg)  Copper_(mg)  Manganese_(mg)  Vit_C_(mg)  \
0        643.0       0.09        0.000           0.000         0.0   
1        659.0       0.05        0.016           0.004         0.0   
2          2.0       0.01        0.001           0.000         0.0   

   Thiamin_(mg)  Riboflavin_(mg)  Niacin_(mg)  Vit_B6_(mg)  Vit_E_(mg)  \
0         0.005            0.034        0.042        0.003        2.32   
1         0.005            0.034        0.042        0.003        2.32   
2         0.001            0.005        0.003        0.001        2.80   

   Cholestrl_(mg)  
0           215.0  
1           219.0  
2           256.0  


I can also do arithmetic to columns.

In [31]:
# arithmetic as series
div_1000 = food_info["Iron_(mg)"] / 1000
add_100 = food_info["Iron_(mg)"] + 100
sub_100 = food_info["Iron_(mg)"] - 100
mult_2 = food_info["Iron_(mg)"]*2
# arithmetic as DataFrames
sodium_grams = food_info[['Sodium_(mg)']] / 1000
sugar_milligrams = food_info[['Sugar_Tot_(g)']] * 1000
print(sodium_grams)

      Sodium_(mg)
0           0.643
1           0.659
2           0.002
3           1.146
4           0.560
5           0.629
6           0.842
7           0.690
8           0.644
9           0.700
10          0.604
11          0.364
12          0.344
13          0.372
14          0.308
15          0.406
16          0.365
17          0.812
18          0.917
19          0.800
20          0.600
21          0.819
22          0.714
23          0.800
24          0.600
25          0.627
26          0.710
27          0.619
28          0.682
29          0.628
...           ...
8588        0.002
8589        0.002
8590        0.007
8591        0.564
8592        0.464
8593        0.490
8594        0.001
8595        0.199
8596        0.297
8597        0.016
8598        0.486
8599        0.000
8600        0.002
8601        1.297
8602        1.435
8603        2.838
8604        0.010
8605        0.002
8606        0.012
8607        0.000
8608        3.326
8609        1.765
8610        3.750
8611      

I can also do column to column transformations.

In [33]:
# multiplying/dividing columns as series
water_energy = food_info["Water_(g)"] * food_info["Energ_Kcal"]
grams_of_protein_per_gram_of_water = food_info['Protein_(g)']/ food_info['Water_(g)']
# adding columns as series
milligrams_of_calcium_and_iron = food_info['Iron_(mg)'] + food_info['Calcium_(mg)']
print(water_energy)

0       11378.79
1       11378.79
2         210.24
3       14970.73
4       15251.81
5       16172.28
6       15540.00
7       14769.28
8       15062.60
9       14570.55
10      15050.80
11       7819.42
12       7725.08
13       5832.72
14       6580.44
15       5938.56
16      18618.48
17      14836.92
18      14578.08
19      14750.88
20       6263.04
21      14759.76
22      13707.47
23      15833.34
24      15296.73
25      15003.00
26      15384.84
27      13660.12
28      13707.54
29      15371.36
          ...   
8588      778.00
8589     6979.70
8590     5650.80
8591      604.50
8592     2005.00
8593     1201.20
8594     5956.80
8595    10668.40
8596     5466.50
8597    13972.00
8598    14929.60
8599        0.00
8600     6668.10
8601    14263.50
8602     2871.00
8603     1495.20
8604     5232.80
8605     9487.00
8606     9893.46
8607     8104.88
8608     2482.00
8609     3650.40
8610     2394.00
8611     3034.00
8612     5978.70
8613    13115.00
8614     7797.75
8615     6994.

Now, lets say I want to build a (hypothetical) basic nutritional index for people who want to eat high-protein, low-fat foods. I'll be using the protein and lipid column, and the following formula to 'score' each food in our data set:
 - score = (2 x protein) - (0.75 x lipids)

In [38]:
weighted_protein = food_info['Protein_(g)']* 2
weighted_fat = food_info['Lipid_Tot_(g)'] * 0.75
initial_rating = weighted_protein - weighted_fat
print(initial_rating)

0       -59.1325
1       -59.1325
2       -74.0500
3        21.2450
4        24.2200
5        20.7400
6        21.4050
7        28.4600
8        22.7150
9        23.7900
10       23.4375
11       19.0150
12       18.4925
13       20.4625
14       19.1975
15       24.0150
16      -13.8200
17       29.1300
18       12.4600
19       27.8450
20       -2.8325
21       29.3000
22       35.3650
23       19.6625
24       26.2500
25       27.5775
26       24.7200
27       36.5800
28       34.3700
29       24.2900
          ...   
8588     12.5250
8589      1.8500
8590      0.4500
8591      6.0000
8592     10.1450
8593     18.0750
8594      0.8500
8595     52.9750
8596      3.1750
8597     42.1750
8598    -58.3500
8599    -75.0000
8600      0.0000
8601     -6.4650
8602     21.0250
8603      8.8000
8604      0.0000
8605      0.5775
8606      0.6700
8607      0.1525
8608     17.9100
8609      3.1250
8610      0.9450
8611    148.9325
8612     32.5750
8613     18.1750
8614     40.4500
8615      0.00

Right now, my data isn't normalized. The range of values varies greatly - e.g kilo-calories, mg, g. etc. So I want to first rescale my data using the rescaling formula:
 - rescaled_column = (original_column - min_value_in_column) / (max_value_in_column - min_value_in_column)

In [41]:
# creating rescaling function
def rescaling(column_name):
    rescaled = (food_info[column_name] - food_info[column_name].min()) / (food_info[column_name].max() - food_info[column_name].min())
    return rescaled

In [47]:
normalized_protein = rescaling('Protein_(g)')
normalized_lipids = rescaling('Lipid_Tot_(g)')
normalized_water = rescaling('Water_(g)')
normalized_carbs = rescaling('Carbohydrt_(g)')

In [52]:
# adding the normalized columns to the dataframe
food_info['Normalized_Protein'] = normalized_protein
food_info['Normalized_Fat'] = normalized_lipids
food_info['Normalized_Carbs'] = normalized_carbs

In [53]:
food_info.head(1)

NDB_No         Shrt_Desc  Water_(g)  Energ_Kcal  Protein_(g)  \
0    1001  BUTTER WITH SALT      15.87         717     0.009624   

   Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  Sugar_Tot_(g)  \
0          81.11     2.11            0.06           0.0           0.06   

         ...         Vit_D_IU  Vit_K_(mcg)  FA_Sat_(g)  FA_Mono_(g)  \
0        ...             60.0          7.0      51.368       21.021   

   FA_Poly_(g)  Cholestrl_(mg)  Normalized_protein  Normalized_Protein  \
0        3.043           215.0            0.009624            0.009624   

   Normalized_Fat  Normalized_Carbs  
0          0.8111            0.0006  

[1 rows x 40 columns]

Now, im going to use the normalized protein and normalized fat columns to redo my nutritional index calculation. I will store those calculations as a new column in the DataFrame - 'Norm_Nutr_Index'.

In [54]:
food_info['Norm_Nutr_Index'] = (2 * food_info['Normalized_Protein']) - (0.75 * food_info['Normalized_Fat'])

In [55]:
food_info.head(3)

NDB_No                 Shrt_Desc  Water_(g)  Energ_Kcal  Protein_(g)  \
0    1001          BUTTER WITH SALT      15.87         717     0.009624   
1    1002  BUTTER WHIPPED WITH SALT      15.87         717     0.009624   
2    1003      BUTTER OIL ANHYDROUS       0.24         876     0.003170   

   Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  Sugar_Tot_(g)  \
0          81.11     2.11            0.06           0.0           0.06   
1          81.11     2.11            0.06           0.0           0.06   
2          99.48     0.00            0.00           0.0           0.00   

        ...         Vit_K_(mcg)  FA_Sat_(g)  FA_Mono_(g)  FA_Poly_(g)  \
0       ...                 7.0      51.368       21.021        3.043   
1       ...                 7.0      50.489       23.426        3.012   
2       ...                 8.6      61.924       28.732        3.694   

   Cholestrl_(mg)  Normalized_protein  Normalized_Protein  Normalized_Fat  \
0           215.0            0.009624            0.009624          0.8111   
1           219.0            0.009624            0.009624          0.8111   
2           256.0            0.003170            0.003170          0.9948   

   Normalized_Carbs  Norm_Nutr_Index  
0            0.0006        -0.589077  
1            0.0006        -0.589077  
2            0.0000        -0.739759  

[3 rows x 41 columns]

I want to sort my 'Norm_Nutr_Index' so that we can see the scores in descending order.

In [56]:
food_info.sort_values('Norm_Nutr_Index', inplace=True, ascending=False)
food_info.head(3)

NDB_No                                 Shrt_Desc  Water_(g)  Energ_Kcal  \
4991   16423  SOY PROT ISOLATE K TYPE CRUDE PROT BASIS       4.98         321   
6155   19177                  GELATINS DRY PDR UNSWTND      13.00         335   
216     1258    EGG WHITE DRIED STABILIZED GLUCOSE RED       6.53         362   

      Protein_(g)  Lipid_Tot_(g)  Ash_(g)  Carbohydrt_(g)  Fiber_TD_(g)  \
4991     1.000000           0.53     3.58            2.59           2.0   
6155     0.969203           0.10     1.30            0.00           0.0   
216      0.958220           0.48     3.63            4.72           0.0   

      Sugar_Tot_(g)       ...         Vit_K_(mcg)  FA_Sat_(g)  FA_Mono_(g)  \
4991            0.0       ...                 0.0       0.066        0.101   
6155            0.0       ...                 0.0       0.070        0.060   
216             0.0       ...                 0.0       0.147        0.173   

      FA_Poly_(g)  Cholestrl_(mg)  Normalized_protein  Normalized_Protein  \
4991        0.258             0.0            1.000000            1.000000   
6155        0.010             0.0            0.969203            0.969203   
216         0.070            20.0            0.958220            0.958220   

      Normalized_Fat  Normalized_Carbs  Norm_Nutr_Index  
4991          0.0053            0.0259         1.996025  
6155          0.0010            0.0000         1.937656  
216           0.0048            0.0472         1.912840  

[3 rows x 41 columns]

It looks like NBD_No 4991 - which is a soy protein isolate - has the highest score.